File: Atelier/Atelier/Models/ShoppableProduct.swift
import Foundation

struct ShoppableProduct: Identifiable, Equatable, Hashable {
    let id: String
    let roomType: RoomType
    let category: String
    let title: String
    let price: Decimal?
    let currency: String
    let retailer: String
    let affiliateURL: URL
    let productURL: URL
    let imageURL: URL
    let width: Double?
    let depth: Double?
    let height: Double?
    let dimensionUnit: String
    let color: String
    let material: String
    let styleTags: [String]
    let visualDescription: String
    let notes: String

    var displayCategory: String {
        category
            .replacingOccurrences(of: "_", with: " ")
            .capitalized
    }

    var priceText: String {
        guard let price else { return "Verify price" }

        let formatter = NumberFormatter()
        formatter.numberStyle = .currency
        formatter.currencyCode = currency
        formatter.maximumFractionDigits = 2
        return formatter.string(from: price as NSDecimalNumber) ?? "$\(price)"
    }

    var shortTitle: String {
        let pieces = title.components(separatedBy: ",")
        return pieces.first?.trimmingCharacters(in: .whitespacesAndNewlines) ?? title
    }
}

enum RoomType: String, Equatable, Hashable {
    case livingRoom = "living_room"
    case bedroom

    var displayName: String {
        switch self {
        case .livingRoom:
            return "Living room"
        case .bedroom:
            return "Bedroom"
        }
    }
}

File: android/app/src/main/java/com/atelier/android/core/model/ShoppableProduct.kt
package com.atelier.android.core.model

import java.math.BigDecimal
import java.text.NumberFormat
import java.util.Currency
import java.util.Locale

enum class RoomType(val rawValue: String, val displayName: String) {
    LivingRoom("living_room", "Living room"),
    Bedroom("bedroom", "Bedroom"),
}

data class ShoppableProduct(
    val id: String,
    val roomType: RoomType,
    val category: String,
    val title: String,
    val price: BigDecimal?,
    val currencyCode: String,
    val retailer: String,
    val affiliateUrl: String,
    val productUrl: String,
    val imageUrl: String,
    val width: Double?,
    val depth: Double?,
    val height: Double?,
    val dimensionUnit: String,
    val color: String,
    val material: String,
    val styleTags: List<String>,
    val visualDescription: String,
    val notes: String,
) {
    val displayCategory: String
        get() = category.replace("_", " ").replaceFirstChar { it.uppercase() }

    val priceText: String
        get() {
            if (price == null) return "Verify price"
            return runCatching {
                NumberFormat.getCurrencyInstance(Locale.US).apply {
                    this.currency = Currency.getInstance(currencyCode)
                    maximumFractionDigits = 2
                }.format(price)
            }.getOrDefault("$price")
        }

    val shortTitle: String
        get() = title.split(",").firstOrNull()?.trim().orEmpty().ifBlank { title }
}
